pkg/epp/handlers/response.go

```diff
@@ -132,22 +134,65 @@ func (s *Server) HandleResponseBody(
 ) (*extProcPb.ProcessingResponse, error) {
 	logger := log.FromContext(ctx)
 	loggerVerbose := logger.V(logutil.VERBOSE)
-	loggerVerbose.Info("Processing HandleResponseBody")
 	body := req.Request.(*extProcPb.ProcessingRequest_ResponseBody)
 
-	res := Response{}
-	if err := json.Unmarshal(body.ResponseBody.Body, &res); err != nil {
-		return nil, errutil.Error{Code: errutil.Internal, Msg: fmt.Sprintf("unmarshaling response body: %v", err)}
+	if reqCtx.Streaming {
+		logger.V(logutil.DEBUG).Info("Processing HandleResponseBody")
+
+		responseText := string(body.ResponseBody.Body)
+		// Example message if "stream_options": {"include_usage": "true"} is included in the request:
+		// data: {"id":"...","object":"text_completion","created":1739400043,"model":"tweet-summary-0","choices":[],
+		// "usage":{"prompt_tokens":7,"total_tokens":17,"completion_tokens":10}}
+		//
+		// data: [DONE]
+		// Noticed that vLLM returns two entries in one response.
+		// We need to strip the `data:` prefix and next Data: [DONE] from the message to fetch response data.
+		//
+		// If include_usage is not included in the request, `data: [DONE]` is returned separately, which
+		// indicates end of streaming.
+		if strings.Contains(responseText, "data: [DONE]") {
+			response := Response{}
+
+			if reqCtx.StreamingIncludeUsage {
+
+				re := regexp.MustCompile(`\{.*(?:\{.*\}|[^\{]*)\}`) // match for JSON object
+				match := re.FindString(responseText)
+				if match == "" {
+					return nil, errutil.Error{Code: errutil.ModelServerError, Msg: fmt.Sprintf("model server returned invalid response: %v", responseText)}
```

Comment:
Since the response format is a bit loosely defined at this point, and we are doing complex parsing here which can introduce bugs, I think we should safely fall back to not recording the metrics, instead of failing the request. This comment applies to all cases where we return an error, if it's not absolutely necessary.

Reply:
Done.